src/NUnitFramework/framework/Assert.Exceptions.Async.cs

```diff
@@ -0,0 +1,281 @@
+        }
+
+        /// <summary>
+        /// Verifies that an async delegate throws an exception of a certain Type
+        /// or one derived from it when called and returns it.
+        /// </summary>
+        /// <param name="code">A TestDelegate</param>
+        public static Task<TActual> CatchAsync<TActual>(AsyncTestDelegate code) where TActual : System.Exception
+        {
+            return Cast<TActual>(ThrowsAsync(new InstanceOfTypeConstraint(typeof (TActual)), code));
+        }
+
+        #endregion
+
+        #region DoesNotThrowAsync
+
+        /// <summary>
+        /// Verifies that an async delegate does not throw an exception
+        /// </summary>
+        /// <param name="code">A TestDelegate</param>
+        /// <param name="message">The message that will be displayed on failure</param>
+        /// <param name="args">Arguments to be used in formatting the message</param>
+        public static Task DoesNotThrowAsync(AsyncTestDelegate code, string message, params object[] args)
+        {
+            Exception caughtException;
+            try
+            {
+                var task = code();
+                return task.ContinueWith(t =>
+                {
+                    // Case where code() either
+                    // - succeeded (synchronously or asynchronously)
+                    // - failed asynchronously
+                    var ex = t.IsFaulted ? t.Exception.InnerException : null;
+                    Assert.That(ex, new ThrowsNothingConstraint(), message, args);
+                    return ex;
+                });
+            }
+            // Case where code() failed synchronously
+            catch (Exception ex)
+            {
+                caughtException = ex;
+            }
+
+            Assert.That(caughtException, new ThrowsNothingConstraint(), message, args);
+
+            return TaskFromResult(caughtException);
+        }
+        /// <summary>
+        /// Verifies that an async delegate does not throw an exception.
+        /// </summary>
+        /// <param name="code">A TestDelegate</param>
+        public static Task DoesNotThrowAsync(AsyncTestDelegate code)
+        {
+            return DoesNotThrowAsync(code, string.Empty, null);
+        }
+
+        #endregion
+
+        #region Helpers
+
+        private static Task<T> TaskFromResult<T>(T value)
+        {
+            #if NET_4_5 || PORTABLE
+
+            return Task.FromResult(value);
+
+            #else
+
+            var tcs = new TaskCompletionSource<T>();
+            tcs.SetResult(value);
+            return tcs.Task;
+
+            #endif
+        }
+
+        private static Task<TActual> Cast<TActual>(Task<Exception> task) where TActual : Exception
+        {
+            return task
+                .ContinueWith(t =>
+                {
+                    if (t.IsFaulted)
+                        ExceptionHelper.Rethrow(t.Exception.InnerException);
```

Comment:
Need to rethrow the inner exception explicitly, otherwise it's the `AggregateException` that is rethrown